hyperspy/signal.py

```diff
@@ -4039,39 +4057,132 @@ def is_rgb(self):
+        >>> marker = hs.markers.point((10, 30, 50), (30, 50, 60), color='red')
+        >>> s.add_marker(marker, permanent=True,
+        >>>        plot_marker=False, plot_signal=False)
+        >>> s.plot(plot_markers=True) #doctest: +SKIP
+
+        Removing a permanent marker
+        >>> s = hs.signals.Signal2D(np.random.randint(10, size=(100, 100)))
+        >>> marker = hs.markers.point(10, 60, color='red')
+        >>> marker.name = "point_marker"
+        >>> s.add_marker(marker, permanent=True)
+        >>> del s.metadata.Markers.point_marker
         """
-        if self._plot is None:
-            self.plot()
-        if plot_on_signal:
-            self._plot.signal_plot.add_marker(marker)
-        else:
-            self._plot.navigator_plot.add_marker(marker)
-        if plot_marker:
+        marker_data_shape = marker._get_data_shape()
+        if (not (len(marker_data_shape) == 0)) and (
+                marker_data_shape != self.axes_manager.navigation_shape):
+            raise ValueError(
+                    "Navigation shape of the marker must be 0 or the "
+                    "same navigation shape as this signal.")
+        if (marker.signal is not None) and (marker.signal is not self):
+            raise ValueError("Markers can not be added to several signals")
+        marker._plot_on_signal = plot_on_signal
+        marker._plot_marker = plot_marker
+        if plot_signal:
+            if self._plot is None:
+                self.plot()
+        if marker._plot_marker:
+            if self._plot is None:
+                self.plot()
+            if marker._plot_on_signal:
+                self._plot.signal_plot.add_marker(marker)
+            else:
+                if self._plot.navigator_plot is None:
+                    self.plot()
+                self._plot.navigator_plot.add_marker(marker)
             marker.plot()
+        if permanent:
+            if not self.metadata.has_item('Markers'):
+                self.metadata.add_node('Markers')
+            for marker_tuple in list(self.metadata.Markers):
+                if marker is marker_tuple[1]:
+                    raise ValueError("Marker already added to signal")
+            name_list = list(self.metadata.Markers.as_dictionary().keys())
+            name = marker.name
+            temp_name = name
+            for i in range(1, 100000):
+                if temp_name in name_list:
+                    temp_name = name + str(i)
+                else:
+                    name = temp_name
+                    break
+            marker.name = temp_name
+            self.metadata.Markers[name] = marker
+            marker.signal = self
+
+    def _plot_permanent_markers(self):
+        marker_dict_list = list(self.metadata.Markers.__dict__.values())
+        for marker_dict in marker_dict_list:
+            if marker_dict['_dtb_value_'] is not False:
+                marker = marker_dict['_dtb_value_']
```

Comment:
the key lookup could be done once - either in the loop, or when creating the the list to begin with:
```python
marker_dict_list = [v['_dtb_value_'] for v in self.metadata.Markers.__dict__.values()]
# OR ...
for marker_dict in marker_dict_list:
    marker = marker_dict['_dtb_value_']
    if marker is not False:
        # do things...
```